src/lib/support/Pool.h

```diff
@@ -67,29 +88,88 @@ class StaticAllocatorBitmap : public StaticAllocatorBase
     void * At(size_t index) { return static_cast<uint8_t *>(mElements) + mElementSize * index; }
     size_t IndexOf(void * element);
 
-    using Lambda = bool (*)(void *, void *);
-    bool ForEachActiveObjectInner(void * context, Lambda lambda);
+    bool ForEachActiveObjectInner(void * context, bool lambda(void * context, void * object));
 
 private:
     void * mElements;
     const size_t mElementSize;
     std::atomic<tBitChunkType> * mUsage;
 };
 
+template <class T>
+class PoolCommon
+{
+public:
+    template <typename... Args>
+    void ResetObject(T * element, Args &&... args)
+    {
+        element->~T();
+        new (element) T(std::forward<Args>(args)...);
+    }
+};
+
+template <typename T, typename Function>
+class LambdaProxy
+{
+public:
+    LambdaProxy(Function && function) : mFunction(std::move(function)) {}
+    static bool Call(void * context, void * target)
+    {
+        return static_cast<LambdaProxy *>(context)->mFunction(static_cast<T *>(target));
+    }
+
+private:
+    Function mFunction;
+};
+
+#if CHIP_SYSTEM_CONFIG_POOL_USE_HEAP
+
+struct HeapObjectListNode
+{
+    void Remove()
+    {
+        mNext->mPrev = mPrev;
+        mPrev->mNext = mNext;
+    }
+
+    void * mObject;
+    HeapObjectListNode * mNext;
+    HeapObjectListNode * mPrev;
+};
+
+struct HeapObjectList : HeapObjectListNode
+{
+    HeapObjectList() { mNext = mPrev = this; }
```

Comment:
Nice optimization making the list circular.  This simplified a number of things.